react/test/quiz_test.js

```diff
@@ -1,43 +1,284 @@
-import { Selector } from 'testcafe';
+import { Selector, RequestHook } from 'testcafe';
 import { TARGET, screencap } from './test_utils';
+import { exec } from 'child_process';
+import { writeFileSync } from 'fs';
 
-fixture('quiz clickthrough test')
-    .page(TARGET + '/quiz.html?date=99')
-    // no local storage
-    .beforeEach(async t => {
-        await t.eval(() => {
-            localStorage.clear()
+async function quiz_screencap(t, name) {
+    await t.eval(() => {
+        const elem = document.getElementById("quiz-timer");
+        if (elem) {
+            elem.remove();
+        }
+    });
+    await t.wait(1000);
+    await screencap(t, name);
+}
+
+export class ProxyPersistent extends RequestHook {
+    constructor() {
+        // only use this hook for persistent.urbanstats.org
+        // super("https://persistent.urbanstats.org", 443);
+        super()
+    }
+
+    async onRequest(e) {
+        if (e.requestOptions.hostname == "persistent.urbanstats.org") {
+            e.requestOptions.hostname = "localhost";
+            e.requestOptions.port = 54579;
+            e.requestOptions.protocol = "http:";
+            e.requestOptions.path = e.requestOptions.path.replace("https://persistent.urbanstats.org", "localhost:54579");
+            e.requestOptions.host = "localhost:54579";
+            // console.log(e)
+        }
+    }
+
+    async onResponse(e) {
+    }
+}
+
+async function run_query(query) {
+    // dump given query to a string
+    const command_line = `sqlite3 ../urbanstats-persistent-data/db.sqlite3 "${query}"`;
+    return new Promise((resolve, reject) => {
+        exec(command_line, (err, stdout, stderr) => {
+            if (err || stderr) {
+                console.log(err);
+                console.log(stderr);
+                reject(err);
+            }
+            resolve(stdout);
         });
     });
+}
+
+function juxtastat_table() {
+    return run_query("SELECT user, day, corrects from JuxtaStatIndividualStats");
+}
+
+function quiz_fixture(fix_name, url, new_localstorage, sql_statements) {
+    fixture(fix_name)
+        .page(url)
+        // no local storage
+        .beforeEach(async t => {
+            // create a temporary file
+            const tempfile = "/tmp/quiz_test_" + Math.floor(Math.random() * 1000000) + ".sql";
+            // write the sql statements to the temporary file
+            writeFileSync(tempfile, sql_statements);
+            await new Promise((resolve, reject) => {
```

Comment:
Using `util.promisify` can eliminate this boilerplate.

Reply:
fixed